File: backend/app/schemas/projections.py
from decimal import Decimal

from pydantic import BaseModel

DISCLAIMER = "Projections are hypothetical estimates and are not guaranteed."


class ProjectionRequest(BaseModel):
    """Input values for a compound-growth projection."""

    initial_amount: Decimal
    monthly_contribution: Decimal
    annual_return_rate: Decimal
    years: int


class YearlyBalanceResponse(BaseModel):
    """A projected balance at the end of a year."""

    year: int
    balance: str


class ProjectionResponse(BaseModel):
    """Response for a hypothetical compound-growth projection."""

    projected_balance: str
    total_contributions: str
    estimated_growth: str
    yearly_balances: list[YearlyBalanceResponse]
    disclaimer: str

File: frontend/types/projections.ts
export type ProjectionFormValues = {
  initialInvestment: number;
  monthlyContribution: number;
  annualReturn: number;
  years: number;
};

export type ProjectionFieldName = keyof ProjectionFormValues;

export type ProjectionValidationErrors = Partial<
  Record<ProjectionFieldName, string>
>;

export type ProjectionRequest = {
  initial_amount: string;
  monthly_contribution: string;
  annual_return_rate: string;
  years: number;
};

export type YearlyBalance = {
  year: number;
  balance: string;
};

export type ProjectionResponse = {
  projected_balance: string;
  total_contributions: string;
  estimated_growth: string;
  yearly_balances: YearlyBalance[];
  disclaimer: string;
};

export type ProjectionResultState =
  | {
      status: "idle";
    }
  | {
      status: "loading";
      submittedValues: ProjectionFormValues;
    }
  | {
      status: "ready";
      submittedValues: ProjectionFormValues;
      projection: ProjectionResponse;
    }
  | {
      status: "error";
      submittedValues: ProjectionFormValues;
      message: string;
    };
